src/sql/src/query_model/hir/hir_from_qgm.rs

```diff
@@ -24,23 +31,393 @@ impl From<Model> for HirRelationExpr {
 }
 
 #[derive(Default)]
-struct FromModel;
+struct FromModel {
+    /// Generates [`mz_expr::LocalId`] as an alias for common expressions.
+    id_gen: IdGen,
+    // TODO: Add the current correlation information.
+    // Inspired by FromHir::context_stack. Subject to change.
+    // context_stack: Vec<QuantifierId>,
+    /// Stack of [HirRelationExpr] that have been created through visiting nodes
+    /// of QGM in post-order.
+    converted: Vec<HirRelationExpr>,
+    /// Common expressions that have been given a [`mz_expr::LocalId`], from
+    /// oldest to most recently identified.
+    lets: Vec<(mz_expr::LocalId, HirRelationExpr, mz_repr::RelationType)>,
+    /// Map of ([BoxId]s whose HIR representation has been given a
+    /// [`mz_expr::LocalId`]) -> (position of its HIR representation in `lets`)
+    common_subgraphs: HashMap<BoxId, usize>,
+}
+
+/// Tracks the column ranges in a [HirRelationExpr] in corresponding the
+/// quantifiers in a box.
+struct ColumnMap {
+    quantifier_to_input: HashMap<QuantifierId, usize>,
+    input_mapper: mz_expr::JoinInputMapper,
+}
 
 impl FromModel {
-    #[allow(dead_code, unused_variables, unused_mut)]
-    fn generate(mut self, model: Model) -> HirRelationExpr {
+    #[allow(dead_code)]
+    fn generate(mut self, mut model: Model) -> HirRelationExpr {
+        // Derive the RelationType of each box.
+        use crate::query_model::attribute::core::{Attribute, RequiredAttributes};
+        let attributes = HashSet::from_iter(std::iter::once(
+            Box::new(BoxRelationType) as Box<dyn Attribute>
+        ));
+        let root = model.top_box;
+        RequiredAttributes::from(attributes).derive(&mut model, root);
+
+        // Convert QGM to HIR.
         let _ = model.try_visit_pre_post(
-            &mut |model, box_id| -> Result<(), QGMError> {
-                println!("pre-visit");
+            &mut |_, _| -> Result<(), QGMError> {
+                // TODO: add to the context stack
                 Ok(())
             },
             &mut |model, box_id| -> Result<(), QGMError> {
-                println!("post-visit");
+                self.convert_subgraph(model, box_id);
                 Ok(())
             },
         );
```

Comment:
nit: add

```rust
debug_assert!(self.converted.len() == 1);
```

here to indicate that we expect a singleton `converted` stack after all boxes have been converted.